File: src/plugin_api/nominatim.rs
//! Nominatim API client — forward and reverse geocoding.
//! https://nominatim.openstreetmap.org/

use log::debug;

use crate::geo::LonLat;
use crate::shared::http::HttpClient;
use crate::shared::http::url::urlencoded;

const BASE_URL: &str = "https://nominatim.openstreetmap.org";

/// Forward geocoding result.
#[derive(Debug, Clone)]
pub struct SearchResult {
    pub name: String,
    pub location: LonLat,
}

/// Reverse geocoding result.
#[derive(Debug, Clone)]
pub struct PlaceInfo {
    pub display_name: String,
    pub city: Option<String>,
    pub country: Option<String>,
}

pub struct NominatimClient {
    http: HttpClient,
}

impl Default for NominatimClient {
    fn default() -> Self {
        Self::new()
    }
}

impl NominatimClient {
    pub fn new() -> Self {
        Self {
            http: HttpClient::new("nominatim"),
        }
    }

    /// Forward geocoding: place name → coordinates.
    pub fn search(&self, query: &str) -> Vec<SearchResult> {
        let url = format!(
            "{}/search?q={}&format=json&limit=5",
            BASE_URL,
            urlencoded(query),
        );
        debug!("nominatim: search {}", url);

        let json = match self.http.get_json::<Vec<serde_json::Value>>(&url) {
            Ok(v) => v,
            Err(e) => {
                log::warn!("nominatim: search fetch failed for \"{}\": {}", query, e);
                return Vec::new();
            }
        };

        json.iter()
            .filter_map(|item| {
                let lat: f64 = item.get("lat")?.as_str()?.parse().ok()?;
                let lon: f64 = item.get("lon")?.as_str()?.parse().ok()?;
                let name = item.get("display_name")?.as_str()?.to_string();
                Some(SearchResult {
                    name,
                    location: LonLat { lon, lat },
                })
            })
            .collect()
    }

    /// Reverse geocoding: coordinates → place info.
    pub fn reverse(&self, lat: f64, lon: f64) -> Option<PlaceInfo> {
        let url = format!(
            "{}/reverse?lat={}&lon={}&format=json&zoom=10",
            BASE_URL, lat, lon,
        );
        debug!("nominatim: reverse {}", url);

        let json: serde_json::Value = self.http.get_json(&url).ok()?;
        let display_name = json.get("display_name")?.as_str()?.to_string();
        let address = json.get("address");
        let city = address
            .and_then(|a| {
                a.get("city")
                    .or_else(|| a.get("town"))
                    .or_else(|| a.get("village"))
            })
            .and_then(|v| v.as_str())
            .map(|s| s.to_string());
        let country = address
            .and_then(|a| a.get("country"))
            .and_then(|v| v.as_str())
            .map(|s| s.to_string());

        Some(PlaceInfo {
            display_name,
            city,
            country,
        })
    }
}

File: ttymap-app/runtime/lua/plugin/search/nominatim.lua
-- search.nominatim — Nominatim forward-geocoding REST client.
--
-- Source: https://nominatim.openstreetmap.org/search
-- Free public endpoint; debounce keeps it from being hammered while
-- the user types.

local M = {}

local SEARCH_URL = "https://nominatim.openstreetmap.org/search"
local LIMIT = 5

function M.url(query)
    return string.format("%s?q=%s&format=json&limit=%d",
        SEARCH_URL, ttymap.http:url_encode(query), LIMIT)
end

function M.parse(payload)
    local out = {}
    if type(payload) ~= "table" then return out end
    for _, item in ipairs(payload) do
        if type(item) == "table"
            and type(item.display_name) == "string"
            and type(item.lat) == "string"
            and type(item.lon) == "string" then
            local lat = tonumber(item.lat)
            local lon = tonumber(item.lon)
            if lat and lon then
                table.insert(out, {
                    name = item.display_name,
                    lon = lon,
                    lat = lat,
                })
            end
        end
    end
    return out
end

return M
